libs/cuda_support/include/hpx/cuda_support/cublas_executor.hpp

```diff
@@ -0,0 +1,252 @@
+                return "CUBLAS_STATUS_ALLOC_FAILED";
+            case CUBLAS_STATUS_INVALID_VALUE:
+                return "CUBLAS_STATUS_INVALID_VALUE";
+            case CUBLAS_STATUS_ARCH_MISMATCH:
+                return "CUBLAS_STATUS_ARCH_MISMATCH";
+            case CUBLAS_STATUS_MAPPING_ERROR:
+                return "CUBLAS_STATUS_MAPPING_ERROR";
+            case CUBLAS_STATUS_EXECUTION_FAILED:
+                return "CUBLAS_STATUS_EXECUTION_FAILED";
+            case CUBLAS_STATUS_INTERNAL_ERROR:
+                return "CUBLAS_STATUS_INTERNAL_ERROR";
+            case CUBLAS_STATUS_NOT_SUPPORTED:
+                return "CUBLAS_STATUS_NOT_SUPPORTED";
+            case CUBLAS_STATUS_LICENSE_ERROR:
+                return "CUBLAS_STATUS_LICENSE_ERROR";
+            }
+            return "<unknown>";
+        }
+    }    // namespace detail
+
+    // -------------------------------------------------------------------------
+    // exception type for failed launch of cuda functions
+    struct HPX_EXPORT cublas_exception : hpx::exception
+    {
+        cublas_exception(const std::string& msg, cublasStatus_t err)
+          : hpx::exception(hpx::bad_function_call, msg)
+          , err_(err)
+        {
+        }
+        cublasStatus_t get_cublas_errorcode()
+        {
+            return err_;
+        }
+
+    protected:
+        cublasStatus_t err_;
+    };
+
+    inline cublasStatus_t check_cublas_error(cublasStatus_t err)
+    {
+        if (err != CUBLAS_STATUS_SUCCESS)
+        {
+            std::stringstream temp;
+            temp << "cublas function returned error code :"
+                 << detail::_cublasGetErrorString(err);
+            throw cublas_exception(temp.str(), err);
+        }
+        return err;
+    }
+
+    namespace detail {
+        // specialization for return type of cublasStatus_t
+        template <typename... Args>
+        struct dispatch_helper<cublasStatus_t, Args...>
+        {
+            inline cublasStatus_t operator()(
+                cublasStatus_t (*f)(Args...), Args... args)
+            {
+                cublasStatus_t err = f(args...);
+                return check_cublas_error(err);
+            }
+        };
+
+    }    // namespace detail
+
+    // -------------------------------------------------------------------------
+    // a simple cublas wrapper helper object that can be used to synchronize
+    // cublas calls with an hpx future.
+    // -------------------------------------------------------------------------
+    struct cublas_executor : cuda_executor
+    {
+        // construct a cublas stream
+        cublas_executor(std::size_t device, bool event_mode = false)
+          : hpx::cuda::cuda_executor(device, event_mode)
+        {
+            detail::cub_debug.debug(
+                debug::str<>("cublas_executor"), "event mode", event_mode);
+
+            handle_ = 0;
+            check_cublas_error(cublasCreate(&handle_));
```

Comment:
CUBLAS handles are an expensive resource, their destruction may cause implicit device synchronization. Have you perhaps considered making a pool of CUBLAS handles from which cublas_executor can get it's CUBLAS handles without creating and destroying them?

https://docs.nvidia.com/cuda/cublas/index.html#cublas-context

Reply:
I ad assumed that the cublas executor would be reasonably long lived - ideally created at startup, but the issue of managing many streams via many executors (and many cublas handles) has not yet been addressed.
I suggest we merge this PR without a handle pool and add support for that later. I've got a pool for the events in place and it ought to be straightforward to add an other pool, but it's worth it's own PR I think.

Reply:
My only concern with managing the lifetime of the CUBLAS handle within the executor is when passing it by value within a closure of a lambda:

```
some_fut.then([cublas_ex, ...] {
    return async(cublas_ex, ....).get() # some cuda call
});
```

this might incur some unwanted synchronizations and may be quite expensive. If the executor is long-lived, perhaps then it would be better to just pass a reference within the closure. What do you think?

Reply:
I thought I  had made the executor non-copyable to prevent that. maybe I accidentally removed that code. The idea is to make it move only and use it by reference and keep it alive until no longer needed.

Reply:
The executors are non copyable, so your lambda would be disallowed. I hope that is satisfactory (for now)